Update the root readme to include clear acquisition and installation details. (#1723)
Azure/azure-sdk-for-cpp: README.md

This repository is for active development of the Azure SDK for C++. For consumers of the SDK we recommend visiting our [developer docs](https://azure.github.io/azure-sdk-for-cpp).

## Getting started

For the best development experience, we recommend developers use [CMake projects in Visual Studio](https://docs.microsoft.com/cpp/build/cmake-projects-in-visual-studio?view=vs-2019) to view and build the source code together with its dependencies. You can also use any other text editor of your choice, such as [VS Code](https://code.visualstudio.com/), along with the command line for building your application with the SDK.

You can find additional information for specific libraries by navigating to the appropriate folder in the `/sdk` directory. See the **README.md** file located in the library's project folder, for example, the [Azure Storage client library](https://github.com/Azure/azure-sdk-for-cpp/tree/master/sdk/storage#azure-storage-client-library-for-c).

For API reference docs, tutorials, samples, quick starts, and other documentation, go to [Azure SDK for C++ Developer Docs](https://azure.github.io/azure-sdk-for-cpp).

### Download & Install the SDK

> The easiest way to acquire the C++ SDK is leveraging vcpkg package manager. You will need to install [Git](https://git-scm.com/downloads) before getting started.

Author: @RickWinter I removed the link to the vcpkg repo from here, since I link it later below and also removed example commands like `> .\vcpkg\vcpkg install [packages to install]`, leaving just the concrete ones.

Let me know if this is what you had in mind, and if you have any other suggestions.

Reviewer: Seems reasonable.  We should give enough detail to onboard and unblock the customer while not becoming the vcpkg documentation.

You can reference this [vcpkg Quick Start](https://github.com/microsoft/vcpkg#quick-start-windows) for more details.

### Building your Application

In order to use the SDK installed via vcpkg with CMake, you can use the toolchain file from vcpkg:

```cmd
> cmake -B [build directory] -S . -DCMAKE_TOOLCHAIN_FILE=[path to vcpkg repo]/scripts/buildsystems/vcpkg.cmake
> cmake --build [build directory]
```

The **entry point** for most scenarios when using the SDK will be a top-level client type corresponding to the Azure service you want to talk to. For example, sending requests to blob storage can be done via the `Azure::Storage::Blobs::BlobClient` API.

All the Azure C++ SDK headers needed to be included are located within the `<azure>` folder, with sub-folders corresponding to each service. Similarly, all types and APIs can be found within the `Azure::` namespace. For example, to use functionality form `Azure::Core`, include the following header at the beginning of your application `#include <azure/core.hpp>`.

Here's an example application to help you get started:

```C++
#include <iostream>

> #include <azure/core.hpp>

Reviewer 2: We can remove this if we are not using anything from core.hpp directly.
If you want to keep it, add something like a body stream or context

Author: `RequestFailedException` gets used.

#include <azure/storage/blobs.hpp>

using namespace Azure::Storage::Blobs;

std::string GetConnectionString()
{
  return "get connection string from azure portal/environment";

> }

Reviewer 2: Well, this gives the impression that we are recommending having the connection string in the source code, which is bad.

Since this is an example of how we would do a sample app, can we use `std::getenv()` to recommend always putting connection string on the env?

Author: This isn't really the point of the sample, and I didn't want to encourage use of `getenv()` especially without "is string empty" checks. Doing it properly ends up taking up too much code space. But you bring up a good point.

Maybe I should just leave it without a body and add a comment that the user can read it from an environment variable.

int main()
{
  std::string containerName = "testcontainer";
  std::string blobName = "sample-blob";

  try
  {
    auto client = BlobClient::CreateFromConnectionString(
        GetConnectionString(), containerName, blobName);
  }
  catch (const Azure::Core::RequestFailedException& e)
  {
    std::cout << e.what() << std::endl;
    return 1;
  }
  return 0;
}
```

#### Visual Studio - CMakeSettings.json

> When building your application via Visual Studio, you can create and update a `CMakeSettings.json` file and include the following properties to let Visual Studio know where the packages are installed and which triplet needs to be used:

Reviewer 2: I would be surprised by people working with the json directly instead of the json viewer from Visual Studio. ( No actions for this comment, just sharing with you)

Author: I was considering including a picture like this, later:
![image](https://user-images.githubusercontent.com/6527137/108920768-1a5c8480-75ea-11eb-9076-1a4c35c7046b.png)

The client libraries can be identified by the naming used for their folder, package, and namespace. Each will start with `azure`, followed by the service category, and then the name of the service. For example `azure-storage-blobs`.

For a complete list of available packages, please see the [latest available packages](https://azure.github.io/azure-sdk/releases/latest/#c) page.

> NOTE: If you need to ensure your code is ready for production we strongly recommend using one of the stable, non-beta libraries.

### Vcpkg

The following SDK library releases are available on [Vcpkg](https://github.com/microsoft/vcpkg):

* `azure-core-cpp`
* `azure-identity-cpp`
* `azure-storage-blobs-cpp`
* `azure-storage-common-cpp`
* `azure-storage-files-datalake-cpp`
* `azure-storage-files-shares-cpp`

> > NOTE: In case of getting linker errors when consuming the SDK on Windows, make sure that [Vcpkg triplet](https://vcpkg.readthedocs.io/en/latest/users/triplets/) being consumed matches the [CRT link flags](https://docs.microsoft.com/cpp/build/reference/md-mt-ld-use-run-time-library?view=msvc-160) being set for your app or library build. See also `MSVC_USE_STATIC_CRT` build flag.

Reviewer 3: ```suggestion
> NOTE: In case of getting linker errors when consuming the SDK on Windows, make sure that [vcpkg triplet](https://vcpkg.readthedocs.io/en/latest/users/triplets/) being consumed matches the [CRT link flags](https://docs.microsoft.com/cpp/build/reference/md-mt-ld-use-run-time-library?view=msvc-160) being set for your app or library build. See also `MSVC_USE_STATIC_CRT` build flag.

```

[![Build Status](https://dev.azure.com/azure-sdk/public/_apis/build/status/cpp/cpp%20-%20client%20-%20ci?branchName=master)](https://dev.azure.com/azure-sdk/public/_build/latest?definitionId=1611&branchName=master)

> This repository is for active development of the Azure SDK for C++. For consumers of the SDK we recommend visiting our [public developer docs](https://azure.github.io/azure-sdk-for-cpp).

Reviewer 3: Do we reed to say "public developer docs", or can we just say "developer docs"?

## Getting started

> For the best development experience, we recommend that developers use the [CMake projects in Visual Studio](https://docs.microsoft.com/cpp/build/cmake-projects-in-visual-studio?view=vs-2019) to view and build the source code together with its dependencies. You can also use any other text editor of your choice, such as [VS Code](https://code.visualstudio.com/), along with the command line for building your application with the SDK.

Reviewer: ```suggestion
For the best development experience, we recommend developers use [CMake projects in Visual Studio](https://docs.microsoft.com/cpp/build/cmake-projects-in-visual-studio?view=vs-2019) to view and build the source code together with its dependencies. You can also use any other text editor of your choice, such as [VS Code](https://code.visualstudio.com/), along with the command line for building your application with the SDK.

```

> You can find additional information for specific libraries by navigating to the appropriate folder in the `/sdk` directory. See the **README.md** file located in the library's project folder. For example, [here is the doc for the storage client library](https://github.com/Azure/azure-sdk-for-cpp/tree/master/sdk/storage#azure-storage-client-library-for-c).

Reviewer: ```suggestion
You can find additional information for specific libraries by navigating to the appropriate folder in the `/sdk` directory. See the **README.md** file located in the library's project folder. For example, [Azure Storage client library](https://github.com/Azure/azure-sdk-for-cpp/tree/master/sdk/storage#azure-storage-client-library-for-c).

```

Author: This is what you are suggesting here, which isn't really a sentence
> For example, Azure Storage client library.`

Maybe I can merge it in the previous sentence:
> See the README.md file located in the library's project folder, for example, the Azure Storage client library.

> For API reference docs, tutorials, samples, quick starts, and other documentation, go to [Azure for C++ Developer Docs](https://azure.github.io/azure-sdk-for-cpp).

Reviewer: ```suggestion
For API reference docs, tutorials, samples, quick starts, and other documentation, go to [Azure SDK for C++ Developer Docs](https://azure.github.io/azure-sdk-for-cpp).

```

### Download & Install the SDK

> The easiest way to get the C++ SDK is to leverage [vcpkg](https://github.com/microsoft/vcpkg). You'll need to have [Git](https://git-scm.com/downloads) installed before getting started.

Reviewer: ```suggestion
The easiest way to acquire the C++ SDK is leveraging [vcpkg](https://github.com/microsoft/vcpkg). [Git](https://git-scm.com/downloads) is required before getting started.

```

Author: I had heard feedback in the past not to use a passive "reference manual" style tone (e.g. x is required) and refer to the reader directly in an active voice (e.g. you need x), which tends to be more engaging.

How about:
> You will need to install Git before getting started?

> The easiest way to get the C++ SDK is to leverage [vcpkg](https://github.com/microsoft/vcpkg). You'll need to have [Git](https://git-scm.com/downloads) installed before getting started.

Reviewer 3: ```suggestion
The easiest way to get the C++ SDK is to leverage [Vcpkg](https://github.com/microsoft/vcpkg). You'll need to have [Git](https://git-scm.com/downloads) installed before getting started.

```

Author: Following existing vcpkg docs, I don't think we need to capitalize it if isn't in a title or start of a sentence. For example:
https://github.com/microsoft/vcpkg#vcpkg-as-a-submodule
https://vcpkg.readthedocs.io/en/latest/tool-maintainers/layout/
https://en.wikipedia.org/wiki/Vcpkg
https://docs.microsoft.com/en-us/cpp/build/vcpkg?view=msvc-160&viewFallbackFrom=vs-2019

Reviewer 3: Oh, so it looks like the name is lowercase "vcpkg". Then sure, we don't capitalize it unless it starts the sentence. I think we need another set of case fixes then.

> First clone and bootstrap vcpkg itself. You can install it anywhere on your machine, but **make note** of the directory where you clone the vcpkg repo.

Reviewer 3: This is great, but do we want to go to explanation of how VcPkg works? Do they have a doc that maybe we can link to instead?

Author: They do, but having the end user link back-forth between various doc locations isn't a good experience. There are only ~3 steps or the customer to follow, so in-lining them here helps them get started easily. I have a link below to the vcpkg repo if they need more info.

> First clone and bootstrap vcpkg itself. You can install it anywhere on your machine, but **make note** of the directory where you clone the vcpkg repo.

Reviewer 3: ```suggestion
First clone and bootstrap Vcpkg itself. You can install it anywhere on your machine, but **make note** of the directory where you clone the Vcpkg repo.

```

Author: Leaving as is, lower-case unless title or start of sentence.

```cmd
> git clone https://github.com/microsoft/vcpkg
```

On Windows:

```cmd
> .\vcpkg\bootstrap-vcpkg.bat
```

On Linux:

```sh
> ./vcpkg/bootstrap-vcpkg.sh
```

> To install the libraries for your project, run the following, optionally specifying the triplet:

Reviewer 3: If we go to explain vcpkg basics, I'd specify, what the triplet is: "the triplet is a combination of a, b, and c [link to vcpkg docs]". And in this sentence, "optionally specifying the triplet (such as X, Y etc.)"

Reviewer: I'd avoid too much detail and instead show "Install x64 on Windows" followed by steps.  Have 2 or 3 of the basic exact commands.  Then link to vcpkg for those that want a more in depth explanation.  
We should avoid being documentation for vcpkg.  Also you should wait to provide deep dive links to external sites as each time its an opportunity to "lose" them in the journey.

Author: Agreed, that was my approach with the docs here.

```cmd
> .\vcpkg\vcpkg install [packages to install]
```

> For example, this will install the `x64-windows` triplet. On Windows, not specifying a triplet will default to x86-windows:

Reviewer 3: ```suggestion
For example, this will install packages for the `x64-windows` triplet. On Windows, not specifying a triplet will default to x86-windows:

```

> For example, this will install the `x64-windows` triplet. On Windows, not specifying a triplet will default to x86-windows:

Reviewer 3: "On Windows, not specifying a triplet will default to x86-windows" - are we sure?

Author: Yes. :)

Reviewer 3: I am not questioning you saw it, I meant would this be the case on all customer Windows machines, i.e. is this behavior guaranteed/documented, or is it selected based on some system variables, and/or the default be changed by Vcpkg in the future? Is it possible to rewrite as less certain statement/avoid the statement about Vcpkg defaults?

Author: Yes, I had chatted with vcpkg folks and they mentioned the default is x86 which won't be changed.
If you don't have any specific vcpkg environment variables set (i.e. clean Windows machine), yes, that is the default.

I think it is important to call this out as it can be surprising to folks since they get hard to debug errors when trying to find the package and build their application. I don't see much benefit to try to make a less certain statement.

Here's a doc that mentions it:
https://vcpkg.readthedocs.io/en/latest/users/triplets/#additional-remarks

```cmd
> .\vcpkg\vcpkg install azure-storage-blobs-cpp:x64-windows
```

> See the [list of packages](https://github.com/Azure/azure-sdk-for-cpp#vcpkg) available via vcpkg below. All Azure C++ SDK package names start with `azure-`.

Reviewer 3: ```suggestion
See the [list of packages](https://github.com/Azure/azure-sdk-for-cpp#vcpkg) available via Vcpkg below. All Azure C++ SDK package names start with `azure-`.

```

You can also search for the libraries you need with the `search` subcommand:

```cmd
> .\vcpkg\vcpkg search [search term]
```

For example:

```cmd
> .\vcpkg\vcpkg search azure-
```

Once the library is installed, follow the instructions from the console output to include the library in your `CMake` application. For example, to include `azure-storage-blobs-cpp`, add the following to your `CMakeLists.txt` file:

```CMake
find_package(azure-storage-blobs-cpp CONFIG REQUIRED)
target_link_libraries(myProject PRIVATE Azure::azure-storage-blobs)
```

> > NOTE: All the Azure client libraries take a dependency on `azure-core-cpp` which provides functionality commonly needed by all Azure clients. When you install any client library via vcpkg, it will bring in all the necessary dependencies as well. You don't need to install those individually to get started.

Reviewer 3: ```suggestion
> NOTE: All the Azure client libraries take a dependency on `azure-core-cpp` which provides functionality commonly needed by all Azure clients. When you install any client library via Vcpkg, it will bring in all the necessary dependencies as well. You don't need to install those individually to get started.

```

> You can reference this [vcpkg Quick Start](https://github.com/microsoft/vcpkg#quick-start-windows) for more details.

Reviewer 3: ```suggestion
You can reference this [Vcpkg Quick Start](https://github.com/microsoft/vcpkg#quick-start-windows) for more details.

```

### Building your Application

> In order to use the SDK installed via vcpkg with CMake, you can use the toolchain file from vcpkg:

Reviewer 3: ```suggestion
In order to use the SDK installed via Vcpkg with CMake, you can use the toolchain file from Vcpkg:

```

```cmd

> > cmake -B [build directory] -S . -DCMAKE_TOOLCHAIN_FILE=[path to vcpkg repo]/scripts/buildsystems/vcpkg.cmake

Reviewer 3: ```suggestion
> cmake -B [build directory] -S . -DCMAKE_TOOLCHAIN_FILE=[path to Vcpkg repo]/scripts/buildsystems/vcpkg.cmake

```

> cmake --build [build directory]
```

The **entry point** for most scenarios when using the SDK will be a top-level client type corresponding to the Azure service you want to talk to. For example, sending requests to blob storage can be done via the `Azure::Storage::Blobs::BlobClient` API.

All the Azure C++ SDK headers needed to be included are located within the `<azure>` folder, with sub-folders corresponding to each service. Similarly, all types and APIs can be found within the `Azure::` namespace. For example, to use functionality form `Azure::Core`, include the following header at the beginning of your application `#include <azure/core.hpp>`.

Here's an example application to help you get started:

```C++
#include <azure/core.hpp>
#include <azure/storage/blobs.hpp>

using namespace Azure::Storage::Blobs;

int main()
{
	std::string containerName = "testcontainer";
	std::string blobName = "sample-blob";

	try
	{

> 		auto client = BlobClient::CreateFromConnectionString(

Reviewer 3: Something wrong with indentation?

>             GetConnectionString(), containerName, blobName);

Reviewer 3: If GetConnectionString() is not a part of SDK, but is something that user has to define, do the declaration after the include headers and before "usnig namespace": `std::string GetConnectionString();`

Reviewer 3: Ore, maybe, even put a body:
```
std::string GetConnectionString()
{
    return ...
}
```

	}

> 	catch (Azure::Core::RequestFailedException& e)

Reviewer 3: catch by const ref?

	{

> 		puts(e.what());

Reviewer 3: use `std::cout`, the "more C++ way"?

		return 1;
	}
	return 0;
}

> ```

Author: I avoided including this in https://github.com/Azure/azure-sdk-for-cpp/tree/UpdatedReadMe since it reveals a bit too much that can just be copy/pasted.

#### Visual Studio - CMakeSettings.json

When building your application via Visual Studio, you can create and update a `CMakeSettings.json` file and include the following properties to let Visual Studio know where the packages are installed and which triplet needs to be used:

```json
{
  "configurations": [
    {

>         "cmakeToolchain": "<path to vcpkg repo>/vcpkg/scripts/buildsystems/vcpkg.cmake",

Reviewer 3: ```suggestion
        "cmakeToolchain": "<path to Vcpkg repo>/vcpkg/scripts/buildsystems/vcpkg.cmake",

```

The client libraries can be identified by the naming used for their folder, package, and namespace. Each will start with `azure`, followed by the service category, and then the name of the service. For example `azure-storage-blobs`.

For a complete list of available packages, please see the [latest available packages](https://azure.github.io/azure-sdk/releases/latest/#c) page.

> NOTE: If you need to ensure your code is ready for production we strongly recommend using one of the stable, non-beta libraries.

### Vcpkg

The following SDK library releases are available on [Vcpkg](https://github.com/microsoft/vcpkg):

* `azure-core-cpp`
* `azure-identity-cpp`
* `azure-storage-blobs-cpp`
* `azure-storage-common-cpp`
* `azure-storage-files-datalake-cpp`
* `azure-storage-files-shares-cpp`

> > NOTE: In case of getting linker errors when consuming the SDK on Windows, make sure that [Vcpkg triplet](https://vcpkg.readthedocs.io/en/latest/users/triplets/) being consumed matches the [CRT link flags](https://docs.microsoft.com/cpp/build/reference/md-mt-ld-use-run-time-library?view=msvc-160) being set for your app or library build. See also `MSVC_USE_STATIC_CRT` build flag.

Reviewer 3: ```suggestion
> NOTE: In case of getting linker errors when consuming the SDK on Windows, make sure that [VcPkg triplet](https://vcpkg.readthedocs.io/en/latest/users/triplets/) being consumed matches the [CRT link flags](https://docs.microsoft.com/cpp/build/reference/md-mt-ld-use-run-time-library?view=msvc-160) being set for your app or library build. See also `MSVC_USE_STATIC_CRT` build flag.

```

Author: This was an intentional casing change to be consistent across the board, and with what the actual `Vcpkg` repo does. I didn't find this mixed casing used anywhere. For example, see https://github.com/microsoft/vcpkg#vcpkg-with-visual-studio-cmake-projects

Reviewer 3: Tha's good, then please make sure we don't have occurrences of using "VcPkg" - I think we have paragraph title called that.

Author: Can you be specific if there are any that I missed in this README? I only saw one and fixed it already.
